### PR TITLE
[WIP] Disallow same sponsor and applicant SSN

### PR DESCRIPTION
**Pull request by cvalarida**
Connects to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2255

### CODE REVIEW

File: docs/schemaform/form-config.md [hunk 1]
```diff
@@ -180,22 +180,29 @@ We've also been adding some additional uiSchema functionality not found in the r
...
   'ui:validations': [
-    function (errors, fieldData, pageData, fieldSchema, errorMessages) {
+    /**
+     * Note the difference between the three data parameters:
+     *
+     * @param {any} fieldData The data for the current field being validated
+     * @param {object} pageData The data for all the fields on the current page
+     * @param {object} formData The data for all the fields in every page
+     */
+    function (errors, fieldData, pageData, formData fieldSchema, errorMessages) {
```
Comment (cvalarida) [inline docs/schemaform/form-config.md]: This is the only section that I actually modified in this file; the rest is Atom removing the whitespace at the end of lines.

File: test/edu-benefits/pages/contactInformation.unit.spec.jsx [hunk 2]
```diff
@@ -53,7 +53,7 @@ describe('Edu pages contactInformation', () => {
     const formDOM = findDOMNode(form);
     submitForm(form);
 
-    expect(formDOM.querySelectorAll('.usa-input-error').length).to.equal(6);
+    expect(formDOM.querySelectorAll('.usa-input-error').length).to.equal(5);
```
Comment (cvalarida) [inline test/edu-benefits/pages/contactInformation.unit.spec.jsx]: I'll be honest with you; I have no idea why this worked before. The country is also a required field, but it's pre-populated with 'United States,' so that doesn't make much sense to me. There are 6 required fields, but 1 is filled in by default.

File: src/js/common/schemaform/review/ArrayField.jsx [hunk 3]
```diff
@@ -164,7 +164,7 @@ class ArrayField extends React.Component {
                           ? <h5>New {uiSchema['ui:options'].itemName}</h5>
                           : null}
                       <SchemaForm
-                          data={item}
+                          pageData={item}
```
Comment (cvalarida) [inline src/js/common/schemaform/review/ArrayField.jsx]: These may end up needing `formData` as well, but I'd rather not squash all data to get `formData` here as well...

File: docs/schemaform/form-config.md (hunk 1 above)
Comment (raquelromano-gov) [inline docs/schemaform/form-config.md]: Ha, yeah took me a second to realize that you weren't obsessively going over documentation and removing unneeded whitespace :)

Isn't the formData also flattened whereas the pageData maintains the form config structure? May want to explain that if so.

Looks like you're missing a comma after formData

File: docs/schemaform/form-config.md [hunk 4]
```diff
@@ -180,22 +180,29 @@ We've also been adding some additional uiSchema functionality not found in the r
...
       errors.addError('My error');
     },
     {
-      validator: (errors, fieldData, pageData, fieldSchema, errorMessages, options) => {
+      validator: (errors, fieldData, pageData, formData fieldSchema, errorMessages, options) => {
```
Comment (raquelromano-gov) [inline docs/schemaform/form-config.md]: comma missing here too

File: package.json [hunk 5]
```diff
@@ -17,6 +17,7 @@
     "test:accessibility": "./script/run-nightwatch.sh --env accessibility",
     "test:unit": "mocha --recursive 'test/**/*.unit.spec.js?(x)' ",
     "test:coverage": "NODE_ENV=test nyc --all --reporter=lcov --reporter=text --reporter=html mocha --no-color --recursive 'test/**/*.unit.spec.js?(x)' ",
+    "test:unit:edu": "mocha --recursive 'test/edu-benefits/**/*.unit.spec.js?(x)' ",
```
Comment (raquelromano-gov) [inline package.json]: 👍

File: src/js/common/schemaform/validation.js [hunk 6]
```diff
@@ -98,7 +98,7 @@ export function transformErrors(errors, uiSchema) {
  * should call addError to add the error.
  */
 
-export function uiSchemaValidate(errors, uiSchema, schema, formData, formContext, path = '') {
+export function uiSchemaValidate(errors, uiSchema, schema, pageData, formData, formContext, path = '') {
```
Comment (raquelromano-gov) [inline src/js/common/schemaform/validation.js]: Any interest in adding some parameter documentation here?

File: src/js/common/schemaform/SchemaForm.jsx [hunk 7]
```diff
@@ -92,17 +92,18 @@ class SchemaForm extends React.Component {
...
-      uiSchemaValidate(errors, uiSchema, schema, formData);
+      uiSchemaValidate(errors, uiSchema, schema, pageData, formData);
     }
     return errors;
   }
 
   render() {
     const {
-      data,
+      pageData, // Unfortunately, Form calls this formData, but don't be confused,
```
Comment (raquelromano-gov) [inline src/js/common/schemaform/SchemaForm.jsx]: ugh, why? Can we change it to call it pageData?

File: src/js/common/schemaform/FormPage.jsx [hunk 8]
```diff
@@ -79,17 +79,34 @@ class FormPage extends React.Component {
...
       uiSchema
-    } = this.props.form[route.pageConfig.pageKey];
+    } = form[route.pageConfig.pageKey];
+
+    // Flatten the data from every page to pass to SchemaForm to eventually be
+    //  used in uiSchemaValidate()
+    // I'd rather not have this here, but console.time() tells me it's executing
+    //  locally in 1.49e+12ms on the edu 5495, which is admittedly a relatively
+    //  small form, but it shouldn't be a significant hit at any rate.
+    const formData = Object.keys(form).reduce((carry, pageName) => {
```
Comment (raquelromano-gov) [inline src/js/common/schemaform/FormPage.jsx]: How about renaming `data` above to `pageData` so that there is no `data` variable.

Also, should we add a flatten helper and use it both here and in `validation.js`?

File: src/js/common/schemaform/SchemaForm.jsx (hunk 7 above)
Comment (raquelromano-gov) [inline src/js/common/schemaform/SchemaForm.jsx]: Wait, I'm not seeing where that's true--what exactly are you referring to as Form here?

File: test/edu-benefits/pages/contactInformation.unit.spec.jsx (hunk 2 above)
Comment (raquelromano-gov) [inline test/edu-benefits/pages/contactInformation.unit.spec.jsx]: This passed with 6 and now passes with 5?

File: docs/schemaform/form-config.md (hunk 1 above)
Comment (cvalarida) [inline docs/schemaform/form-config.md]: OCD!
MUST REMOVE ALL TRAILING WHITESPACE!

The `formData` is basically all the pages' `data` contents slapped together in one object. That's all the more it's flattened. `pageData` is just the `data` on the current page, so they both have essentially the same structure. `formData` just has more`.

....bloody commas....should lint these .md files. >_>

File: test/edu-benefits/pages/contactInformation.unit.spec.jsx (hunk 2 above)
Comment (cvalarida) [inline test/edu-benefits/pages/contactInformation.unit.spec.jsx]: Yup. No idea why.

File: src/js/common/schemaform/validation.js (hunk 6 above)
Comment (cvalarida) [inline src/js/common/schemaform/validation.js]: Not even a little! But at this point, we probably should have more than just the schemaform documentation...Trouble is I don't really know what type `errors` is, what `formContext` is, nor what `path` is really used for, so... 😕

File: src/js/common/schemaform/SchemaForm.jsx (hunk 7 above)
Comment (cvalarida) [inline src/js/common/schemaform/SchemaForm.jsx]: The Form from react-jsonschema-form, which means no, we can't rename the prop 😞

File: src/js/common/schemaform/FormPage.jsx (hunk 8 above)
Comment (cvalarida) [inline src/js/common/schemaform/FormPage.jsx]: Ah, a flatten helper might be good. We may end up having to do this elsewhere as well.

File: src/js/common/schemaform/validation.js (hunk 6 above)
Comment (jbalboni) [inline src/js/common/schemaform/validation.js]: I'll add some

File: src/js/edu-benefits/5495/config/form.js [hunk 9]
```diff
@@ -104,6 +104,15 @@ const formConfig = {
               'view:noSSN': {
                 'ui:title': 'I don’t know my sponsor’s Social Security number',
               },
+              veteranSocialSecurityNumber: {
+                'ui:validations': [
+                  (errors, fieldData, pageData, formData) => {
+                    if (fieldData === formData.relativeSocialSecurityNumber) {
+                      errors.addError("Your sponsor's SSN cannot be the same as yours.");
```
Comment (jbalboni) [inline src/js/edu-benefits/5495/config/form.js]: Single quotes

File: src/js/common/schemaform/validation.js [hunk 10]
```diff
@@ -96,9 +96,19 @@ export function transformErrors(errors, uiSchema) {
...
+ * @param {Object} pageData The data for the page the field is in
+ * @param {Object} formData The (flattened) data for the entire form
+ * @param {Object} formContext The formContext object passed to rjsf's Form component.
+ *   Should be removed at some point
+ * @param {String} path The path to the current field relative to the root of the page.
+ *   Used to select the correct field data to validate against
  */
 
-export function uiSchemaValidate(errors, uiSchema, schema, formData, formContext, path = '') {
+export function uiSchemaValidate(errors, uiSchema, schema, pageData, formData, formContext, path = '') {
```
Comment (cvalarida) [inline src/js/common/schemaform/validation.js]: I tried using just `formData` at first, but ran into issues...I don't remember what they were exactly, and it could be that they were unrelated, now that I think about it. I'll give it another shot without `pageData` after I get the tests all fixed up.

File: src/js/edu-benefits/5495/config/form.js (hunk 9 above)
Comment (cvalarida) [inline src/js/edu-benefits/5495/config/form.js]: Lint didn't yell at me...Probably because I've got the single quote as an apostrophe for "sponsor's".

Fair enough.

File: src/js/common/schemaform/validation.js [hunk 11]
```diff
@@ -96,6 +96,16 @@ export function transformErrors(errors, uiSchema) {
  *
  * The function is passed errors, fieldData, formData, and otherData and
```
Comment (cvalarida) [inline src/js/common/schemaform/validation.js]: The function is passed errors, fieldData, pageData, formData, and otherData (wha..?)

File: src/js/common/schemaform/FormPage.jsx (hunk 8 above)
Comment (cvalarida) [inline src/js/common/schemaform/FormPage.jsx]: Schemaform helpers.js has `flattenFormData()` which may work, but I'll want to look into the differences before I commit to that. I'm just writing this here so I don't forget...

File: test/edu-benefits/pages/contactInformation.unit.spec.jsx (hunk 2 above)
Comment (cvalarida) [inline test/edu-benefits/pages/contactInformation.unit.spec.jsx]: I haven't investigated it thoroughly yet, but this is because it was being validated with previous information. I just fixed up a bug where the validation was run with the data before it was changed. That made this go back to 6.

I'm guessing the `data={{}}` sets the country to blank, overriding the default..? That'd cause there to be 6 errors instead of 5.

File: src/js/common/schemaform/helpers.js [hunk 12]
```diff
@@ -141,7 +141,7 @@ export function formatReviewDate(dateString) {
   return undefined;
 }
 export function parseISODate(dateString) {
-  if (dateString) {
+  if (typeof dateString === 'string') {
```
Comment (jbalboni) [inline src/js/common/schemaform/helpers.js]: What's the reason for this change (it's fine, just curious)?

**Review by jbalboni — APPROVED** [review]

File: src/js/common/schemaform/helpers.js (hunk 12 above)
Comment (cvalarida) [inline src/js/common/schemaform/helpers.js]: Oh, this was back when the DefinitionTester passed non-objects as data. An object was passed to this function which, of course, passed this statement, but failed when we tried to `dateString.split()`.

Shouldn't be a problem now, but it just stays there as a sort of safeguard.
